### ISSUE TITLE
CSRF Protection

### ISSUE THREAD

**Issue by Izhaki** [vvo/iron-session#431]
This issue is discussed in length in https://github.com/vvo/iron-session/issues/327, but the title is misleading and the issue is closed.

For all I can tell, the current implementation is vulnerable to CSRF attacks, as described in [here](https://portswigger.net/web-security/csrf). I'm pretty sure https://iron-session-example.vercel.app/ is broken at the moment (can't see a login button and get 500 when I click anywhere on the page), but I can easily demonstrate it.

It isn't quite a rocket science to protect against such attacks. In general you need:

1. Send a CRSF token in a cookie together with the session.
2. Send that token in the request body with each POST request (unless GET routes need protection as well).
3. Verify the token on the server.

Step 2 is what an offending site cannot do - although cookies (your session and CRSF) will be sent to the endpoints (because `SameSite=Lax`), the site cannot read these cookies (only same origin site can). So _not_ sending the CRSF token via cookies protects against such attack.

You can look at the [implementation of next-auth](https://github.com/nextauthjs/next-auth/blob/76bf524e8ec144162bcd0474adba250028f9c4db/src/core/init.ts#L110-L130) for some idea on how this is done.

I may be able to submit a PR.

**Comment by vvo** [comment]
Hey there @Izhaki thanks for the comment and feedback here. I am looking to get better on this subject.

For us to focus on the right things, apart from the link you sent, could you write down an imaginary scenario that would prove that using iron-session could be vulnerable to CSRF attacks? From what I remember, because we use SameSite lax then, unless the attacker manages to create a webpage on your domain your cookies will only be sent when navigating to your website, not when an attacker creates evilwebsite.com with a form making post requests to your website (they will not be sent).

I am just trying to find exactly which scenario we need to cover instead of blindly applying CSRF tokens.

Thanks!

**Comment by vvo** [comment]
This page: https://portswigger.net/web-security/csrf/samesite-cookies summarize the technique used by iron-session to prevent CSRF attacks using SameSite. But it relies on users not implementing sensitive actions behind GET requests, which we cannot control.

So I would be down to implementing CSRF tokens in iron-session if there's a nice way to do it. Let me know if you'd like to tackle it and we can discuss next steps/API design.

Thanks a lot!

**Comment by Izhaki** [comment]
Hi @vvo, as you said and referenced, there are very specific conditions for a CSRF attack happening with `SameSite=Lax`, which is what this library is using.

You can see it happening below, albeit with a GET response.

So to prevent such attacks, the API must ensure that all requests that mutate data are POST ones.

I think, as first step, this should be mentioned in the README somewhere.

I've landed on this repo today, so need a bit of time to try it out. Once I have, I might have a go at this.

Let's leave this issue open for now.

![csrf-iron-session](https://user-images.githubusercontent.com/880132/144131982-b2e2872d-7d7e-4fa6-98a4-eaf0d7bcf4ae.gif)

**Comment by Izhaki** [comment]
I have been in conversations with the `next-auth` authors regarding CSRF today, which made me understand why the idea of adding CSRF protection to this library is a **terrible idea**.

I think one of the main appeals of `iron-session` is that it does one thing and does it well. The fact that it is tightly scoped, and has a tiny API surface, make it maintainable and more inviting for contributors (this is in contrast to `next-auth`).

The issue with implementing CSRF in this library is that it adds complexity for both users and maintainers, whereas CSRF is a cross-cutting concern, and should really be implemented in a separate project (there are a few around, working myself on one).

It turns out that `next-auth` only provide CSRF protection to their own (internal) endpoints; to protect other endpoint is the user's responsibility (which might as well mean using a separate, additional solution). There are quite a few issues there regarding CSRF, which requires solid knowledge of the topic to respond to.

While I do understand that it would be nice to have the CSRF tick out of the box for this library, this is not what this library is for. Further, the library itself does not actually export any routes that need protection (unlike `next-auth`) - it is the examples that need CSRF protection.

I hope I've made my case. Closing.
